### PR TITLE
Introducing PublishOptions to allow users to extend the publish api

### PR DESCRIPTION
**Pull request by andreasohlund**
* This one still doesn't allow headers to be set for events, I'll raise that in a separate pull

* We need to look at SendOptions and decide if there are things on there that should be hidden away as well. Eg. destination shouldn't really be in the face of the users. I'll get another pull going for this

* The extensions are now strongly typed and available on the outgoing context(s)

* We need a way to pass extension context items on to the transports. We'll handle that in a separate pull as well

### CODE REVIEW

**Comment by andreasohlund** [comment]
@Particular/nservicebus please review

**Comment by johnsimons** [comment]
I would also have non typed options in `ExtensionContext`, at the moment the current design makes the consumer have a class per setting, that IMO is a burden, eg:
```c#
sendOptions.GetContext().Set(new MyExtensionData { SomeValue = "I did it!" });
// instead of just
sendOptions.GetContext().Set("SomeValue", "I did it!");
```
Also it won't allow anonymous types, eg:
```c#
sendOptions.GetContext().Set(new {Foo = "sdfsdfdsf"}); // You can set it, but u can't get to me!
```

Thinking more about it, I actually think that having a types api is confusing and error prone, users can override each others settings, eg:
```c#
sendOptions.GetContext().Set("My value");

// and later on some other call

sendOptions.GetContext().Set("This overrides previously set string!");
```

I do prefer the explicitness of the "key/value" pair, so :-1: to have a typed api for this scenario.

File: src/NServiceBus.Core/ISendOnlyBus.cs [hunk 1]
```diff
@@ -10,19 +10,15 @@ public interface ISendOnlyBus : IDisposable
         /// <summary>
         /// Publish the message to subscribers.
         /// </summary>
-        void Publish(object message);
-
-        /// <summary>
-        /// Publish the message to subscribers.
-        /// </summary>
-        void Publish<T>();
+        void Publish(object message,PublishOptions options);
```
Comment (johnsimons) [inline src/NServiceBus.Core/ISendOnlyBus.cs]: Should add parameters doco too

**Comment by johnsimons** [comment]
@andreasohlund this PR is good to go once https://github.com/Particular/NServiceBus/pull/2720#issuecomment-97611200 is addressed

**Comment by andreasohlund** [comment]
>I would also have non typed options in ExtensionContext, at the moment the current design makes the >consumer have a class per setting, that IMO is a burden

I don't agree, that's IMO a small price to pay an makes the code much more explicit. Remember this is a api mostly to be used by us. C# is strongly typed so lets us it?

>Thinking more about it, I actually think that having a types api is confusing and error prone, users can override each others settings

I'd say the opposite, with types we're much less likely to interfere with each others since the type will clash and give a compiler error.

I agree about the string example you give, I propose we just put a `where T : class` contraint on to prevent that to be used.

>I do prefer the explicitness of the "key/value" pair

I don't see how kvp string+ object can be more explicit than a strongly typed DTO?


@SzymonPobiega @SimonCropp @distantcam  we've talked about this in the past, please correct me if I'm out of line here?

**Comment by SzymonPobiega** [comment]
Unfortunately `where T : class` would nothelp for strings but would help for ints and such. But we can plug that by throwing an exception `You can't just add a primitive here`. 

I like the strongly typed API because, well, it is strongly typed. Key/value pair in the best case scenario requires me to add an artificial key (most likely the DTOs qualified name) when I do want to have DTOs. In the worst case there would be an explosion of primitives. Key/value pair also brings problems with key clashes. How do I ensure my extension uses a different key than other? Most likely I we would have to guide users to have some convention (e.g. use name spaces). All that goes away with strongly typed API.

**Comment by andreasohlund** [comment]
After talking to @SimonCropp , @johnsimons and @SzymonPobiega we decided to go with a strongly types api. We now also require the context to be a nested class to prevent strings etc to be used.

Pull has been updated to reflect this.

Once green I will pull this in and rebase the #2689 to make sure the api is still what that pull needs

**Comment by danielmarbach** [comment]
@andreasohlund @SimonCropp @johnsimons @SzymonPobiega I think we are inconsistent how we design our APIs. This PR [just removed](https://github.com/Particular/NServiceBus/pull/2720/files#diff-1eb180a4f6d35cd0c53fde55c1a077b3L428) the the `Publish` overloads on `ISendOnlyBus` but introduced [extension methods](https://github.com/Particular/NServiceBus/pull/2720/files#diff-5387bd28da42ddb6bdb12e60c6728ec2R8) for backwards compatibility. BUT when we look at the changes for the `Send` overloads we essentially obsoleted all of them on `ISendOnlyBus` and added them to the `ISendOnlyBusExtensions` class.

```
        [System.ObsoleteAttribute("Please use `Send(object message, SendOptions options)` instead. Will be removed i" +
            "n version 7.0.0.", true)]
        NServiceBus.ICallback Send(NServiceBus.Address address, object message);
        [System.ObsoleteAttribute("Please use `Send<T>(Action<T> messageConstructor, SendOptions options)` instead. " +
            "Will be removed in version 7.0.0.", true)]
        NServiceBus.ICallback Send<T>(NServiceBus.Address address, System.Action<T> messageConstructor);
        [System.ObsoleteAttribute("Please use `Send(object message, SendOptions options)` instead. Will be removed i" +
            "n version 7.0.0.", true)]
        NServiceBus.ICallback Send(string destination, string correlationId, object message);
        [System.ObsoleteAttribute("Please use `Send(object message, SendOptions options)` instead. Will be removed i" +
            "n version 7.0.0.", true)]
        NServiceBus.ICallback Send(NServiceBus.Address address, string correlationId, object message);
        [System.ObsoleteAttribute("Please use `Send<T>(Action<T> messageConstructor, SendOptions options)` instead. " +
            "Will be removed in version 7.0.0.", true)]
        NServiceBus.ICallback Send<T>(string destination, string correlationId, System.Action<T> messageConstructor);
        [System.ObsoleteAttribute("Please use `Send<T>(Action<T> messageConstructor, SendOptions options)` instead. " +
            "Will be removed in version 7.0.0.", true)]
        NServiceBus.ICallback Send<T>(NServiceBus.Address address, string correlationId, System.Action<T> messageConstructor);
```

How do we want to play this? Reintroduce the old `Publish` overloads on `ISendOnlyBus` and deprecate those properly? Thoughts?

**Pull request merged**